README
Bio::Das::Lite
==============


Bio::Das::Lite is an implementation of the BioDas protocol, http://biodas.org/
for the retrieval of biological data from XML sources over HTTP.

Bio::Das::Lite is designed as a lightweight and more forgiving alternative to
the client/retrieval/parsing components of Bio::Das. Bio::Das::Lite itself is
not a drop-in replacement for Bio::Das but it can be subclassed to do so.

NOTE
Results are returned hashed against the query URL which caused them.
This means that sometimes you receive more results than you might expect
(number of sources * number of segments)


INSTALLATION

To install this module type the following:

   perl Build.PL
   ./Build
   ./Build test
   ./Build install


DEPENDENCIES

This module requires these other modules and libraries:

WWW::Curl
HTTP::Response
English
Carp
Readonly

Please note that some elements of HTTP proxy support require that WWW::Curl has been built from a minimum version of the libcurl C library:
  authenticating proxies: 7.19.1
  noproxy (proxy exclusion lists): 7.19.4
If you have an earlier version installed, you will not be able to use these features of Bio::Das::Lite. Regular proxy support should be unaffected.

COPYRIGHT AND LICENCE

Copyright (c) 2007 by Roger Pettett, Genome Research Ltd.
http://www.sanger.ac.uk/

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension Bio::Das::Lite.

2.11 feature.type parsing fix from Nick Langridge

2.10 DAS 1.6 spec, all patches from Andy

2.05 Probably the last release on the DAS 1.53 spec

2.04 Patches to catch out-of-date versions of libcurl.
     Refactored some routines to reduce complexity.

2.01 Bugfix for authenticating proxies.
     No longer appends ? to DAS request URIs without query params.
     Supports links with no descriptions.

2.00 Switched to WWW::Curl::Simple (from LWP::Parallel::UserAgent)

1.61 Minor: Fixed prereq (for CPANTS prereq_matches_use)

1.60 Latest LWP has broken support for "parse_head"
     function. Das::Lite attempts to correct for this by squirting an
     ersatz version into the symbol table under LWP::Protocol, where
     it used to be. A particularly ugly hack!

1.58  Added 'no_proxy' configurable parameter.
      Added 'max_hosts' configurable parameter.
      Supports the X-DAS-Status header.
      Supports score-based stylesheets (see DAS 1.53E spec).
      Fixes for perlcritic.
      Fixed issues of failings tests caused by retired DAS servers.

1.53  kwalitee updates (Test::Pod::Coverage & LICENSE in Makefile.PL)

1.50  added 'sequence' patch for alignments, patch from jt6

1.49  new 'maxbins' parameter for feature requests:

      In recent months more and more das sources appeared that provide
      very condensed value based data. Sometimes it is a feature per
      every 50 basepairs or even a feature per amino acid. It causes
      two major problems:
      1 - amount of XML being transferred is huge
      2 - Ensembl can only do simple merge of the features, i.e just
      pick max or min value - and in some cases das sources providers
      want to use a completely different dataset for a different size
      and resolution of the viewed region.

      To solve these problems we have introduced a new `maxbins`
      parameter to the `features` das request.  Using this parameter
      Ensembl tells all das sources what is the width in pixels of the
      image used to display the features.  Then the das sources smart
      enough to understand the parameter will merge the features
      themselves and for the rest Ensembl will do the work.

1.48  'dna' request support. Minor refactorings and cleanups for perl-critic

1.47  Reworked tests, a couple of fixes for structure handling

1.44  Package namespace changed to Bio::Das::Lite
      -compat tests added for compatibility with deprecated DasLite.pm
      Codebase cleaned up to make Perl::Critic happy
      Protein structure support, patch from rdf
      'sources' request support, patch from rdf

1.41  Protein alignment support, patch from rdf

1.40  Cleanup for Perl::Critic purposes

1.39  Minor typographical tidying up, some more tests

1.38  Minor tweaks to http_proxy - support for http://user:pass@proxy.example.com:nnnn/ format http_proxy

1.33  Minor tweaks to proxy_* methods; Addition of configurable user_agent

1.32  Bugfixes: Cache cleaning for performing multiple queries using persistent objects
                Regex fixes for group notes and links
                Initial support for authenticating proxies (proxy_user, proxy_pass)

1.27  Switched over to CVS version to make things easier for me to track
      Minor revisions. Some more tests
 
0.15  Includes registry support.
      See 'new_from_registry', 'registry' and 'registry_sources' methods.

0.14  Fix for entry_points structure and a new test for it
      Added a new way to use the constructor: my $das = Bio::DasLite->new("http://server/das/source");
      Added a convenience side-effect - give segment data in responses for segments without features in

0.13  Absolutely minor change to include LWP::P::UA max_req connection throttling
      Upgrade strongly recommended

0.12  Reworked all the tests.
      Tweaked basename regex to allow more flexibility in
      specifying services rather than dsns, i.e.
      'dsn' => 'http://das.mysite.com/das'
      rather than
      'dsn' => 'http://das.mysite.com/das/dsn'

0.11  Added a little extra debug, some extra eval traps, improved url processing
      and a couple of new WTSI extensions to the spec

0.10  Fixed dsns DAS request (source basename)

0.09  Fixed broken 'types' request

0.08  Added multi-link, multi-note, multi-group-link and multi-group-note support
      Added X-Forwarded-For HTTP header for IP-based ACLs
      Added passthrough for HTTP header data

0.07  Internal development version

0.06  Fixes for the regex/stack problems in 0.05.
      Added callback support for stylesheet and features calls

0.05  Stylesheet support. Seems to tickle some perl regex/stack bugs
      on certain platforms (x86 Linux, Alpha Tru64). It works fine on my G4.

0.04  Internal version. Added duplication of segment data across
      features in the 'features' call.

0.03  Allowed for calls (namely 'features') with an arrayref of
      segments, i.e. $das->features(["1:1,1000", "2:1,1000"]);
      Minor modifications following some benchmarking and profiling.
      Results are now cached (per Bio::DasLite instance) against
      query URL by default.

0.02  Following feedback, data structures returned by:
      dsn, basename, dsns, entry_points, types, features, sequence
      have now been homogenised across the board. Results are now
      always hashed against the query which was made, regardless
      of how many sources are configured.
      Also added a few more tests including feature-by-id

0.01  Fri Aug 19 16:38:05 2005
	- original version; created by h2xs 1.23 with options
		-cfn Bio::DasLite



t/FileStub.pm
#########
# Author:        rmp
# Last Modified: $Date$ $Author$
# Id:            $Id$
# Source:        $Source: /var/lib/cvsd/cvsroot/Bio-DasLite/Bio-DasLite/t/FileStub.pm,v $
# $HeadURL$
#
package t::FileStub;
use base qw(Bio::Das::Lite);
use strict;
use warnings;
use English qw(-no_match_vars);

our $VERSION = do { my @r = (q$Revision$ =~ /\d+/mxg); sprintf '%d.'.'%03d' x $#r, @r };

sub new {
  my ($class, $ref) = @_;
  $ref ||= {};

  my $self = $class->SUPER::new($ref);

  $self->{'filedata'} = $ref->{'filedata'};
  return $self;
}

sub _fetch {
  my ($self, $url_ref, $headers) = @_;

  open my $fh, q(<), $self->{'filedata'} or die "Cannot open $self->{'filedata'}:[$ERRNO]\n";
  local $RS = undef;
  my $xml  = <$fh>;
  close $fh;

  for my $code_ref (values %{$url_ref}) {
    &{$code_ref}($xml);
  }
  return;
}

1;


t/FileStubCompat.pm
#########
# Author:        rmp
# Last Modified: $Date$ $Author$
# Id:            $Id$
# Source:        $Source: /var/lib/cvsd/cvsroot/Bio-DasLite/Bio-DasLite/t/FileStubCompat.pm,v $
# $HeadURL$
#
package t::FileStubCompat;
use base qw(Bio::DasLite);
use strict;
use warnings;
use English qw(-no_match_vars);

our $VERSION = do { my @r = (q$Revision$ =~ /\d+/mxg); sprintf '%d.'.'%03d' x $#r, @r };

sub new {
  my ($class, $ref) = @_;
  $ref ||= {};

  my $self = $class->SUPER::new($ref);

  $self->{'filedata'} = $ref->{'filedata'};
  return $self;
}

sub _fetch {
  my ($self, $url_ref, $headers) = @_;

  open my $fh, q(<), $self->{'filedata'} or die "Cannot open $self->{'filedata'}:[$ERRNO]\n";
  local $RS = undef;
  my $xml  = <$fh>;
  close $fh;

  for my $code_ref (values %{$url_ref}) {
    &{$code_ref}($xml);
  }
  return;
}

1;


t/FileStubChunked.pm
#########
# Author:        rmp
# Last Modified: $Date$ $Author$
# Id:            $Id$
# Source:        $Source: /var/lib/cvsd/cvsroot/Bio-DasLite/Bio-DasLite/t/FileStubChunked.pm,v $
# $HeadURL$
#
package t::FileStubChunked;
use strict;
use warnings;
use base qw(t::FileStub);

our $VERSION = do { my @r = (q$Revision$ =~ /\d+/mxg); sprintf '%d.'.'%03d' x $#r, @r };

sub _fetch {
  my ($self, $url_ref, $headers) = @_;

  open my $fh, q(<), $self->{'filedata'};
  while(my $xml = <$fh>) {
    for my $code_ref (values %{$url_ref}) {
      &{$code_ref}($xml);
    }
  }
  close $fh;
  return;
}

1;


Build.PL
#########
# Author:        rmp
# Last Modified: $Date$
# Id:            $Id$
# Source:        $Source$
# $HeadURL$
#
use strict;
use warnings;
use Module::Build;
use lib qw(lib);

my $builder = Module::Build->new(
				 module_name         => 'Bio::Das::Lite',
				 dist_author         => 'Roger Pettett <rmp@psyphi.net>',
				 dist_abstract       => q[Bio::Das::Lite is an implementation of the BioDas protocol, http://biodas.org/],
				 license             => 'perl',
				 requires            => {
							 'WWW::Curl'         => '4.07',
							 'HTTP::Response'    => '1.49',
							 'English'           => '1.01',
							 'Carp'              => '1.02',
							 'Readonly'          => '1.03',
							},
				 recommends          => {
							},
				 build_requires      => {
							 'Test::More'              => '0.8',
							},
				 build_recommends    => {
							 'Test::Distribution'      => '2.00',
							 'Test::Perl::Critic'      => '1.01',
							},
				 sign                => 1,
				);

$builder->create_build_script();

print q(
******************************************************************************
** Some of the tests for this module need internet access                   **
** Also remember to set your http_proxy environment variable if you need to **
******************************************************************************
);


Makefile.PL
#########
# Author:        rmp
# Last Modified: $Date: 2010/02/04 15:01:56 $ $Author: aj5 $
# Id:            $Id: Makefile.PL,v 1.48 2010/02/04 15:01:56 aj5 Exp $
# Source:        $Source: /var/lib/cvsd/cvsroot/Bio-DasLite/Bio-DasLite/Makefile.PL,v $
# $HeadURL$
#
package Makefile;
use strict;
use warnings;
use 5.005;
use ExtUtils::MakeMaker;
use English qw(-no_match_vars);

our $VERSION = do { my @r = (q$Revision: 1.48 $ =~ /\d+/mxg); sprintf '%d.'.'%03d' x $#r, @r };

WriteMakefile(
	      'NAME'          => 'Bio::Das::Lite',
	      'AUTHOR'        => q(Roger Pettett <rpettett@cpan.org>),
	      'ABSTRACT_FROM' => 'lib/Bio/Das/Lite.pm',
	      'VERSION_FROM'  => 'lib/Bio/Das/Lite.pm',
	      'DISTNAME'      => 'Bio-Das-Lite',
	      'PREREQ_PM'     => {
                                  'WWW::Curl'         => '4.07',
				  'HTTP::Response'    => '1.49',
				  'SOAP::Lite'        => '0.69',
				  'English'           => '1.01',
				  'Carp'              => '1.02',
				  'Readonly'          => '1.03',
				 },
	      'dist'          => {
				  COMPRESS => 'gzip',
				  SUFFIX   => 'gz',
				 },
	      'LICENSE'       => 'perl',
	     );

print q(
******************************************************************************
** Some of the tests for this module need internet access                   **
** Also remember to set your http_proxy environment variable if you need to **
******************************************************************************
);

1;
